bootstrap/internal/controllers/rke2config_controller.go

```diff
@@ -225,25 +240,329 @@ func (r *Rke2ConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		return ctrl.Result{}, err
+	}
+	conditions.MarkTrue(scope.Config, bootstrapv1.CertificatesAvailableCondition)
+
+	token, err := r.generateAndStoreToken(ctx, scope)
+	if err != nil {
+		scope.Logger.Error(err, "unable to generate and store an RKE2 server token")
+		return ctrl.Result{}, err
+	}
+	scope.Logger.Info("RKE2 server token generated and stored in Secret!")
+
+	configStruct, files, err := rke2.GenerateInitControlPlaneConfig(
+		rke2.RKE2ServerConfigOpts{
+			ControlPlaneEndpoint: scope.Cluster.Spec.ControlPlaneEndpoint.Host,
+			Token:                token,
+			ServerURL:            "https://" + scope.Cluster.Spec.ControlPlaneEndpoint.Host + ":9345",
+			ServerConfig:         scope.ControlPlane.Spec.ServerConfig,
+			AgentConfig:          scope.Config.Spec.AgentConfig,
+			Ctx:                  ctx,
+			Client:               r.Client,
+		})
+
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	b, err := kubeyaml.Marshal(configStruct)
+	if err != nil {
+
+		return ctrl.Result{}, err
+	}
+	scope.Logger.Info("Server config marshalled successfully")
+
+	initConfigFile := bootstrapv1.File{
+		Path:        rke2.DefaultRKE2ConfigLocation,
+		Content:     string(b),
+		Owner:       "root:root",
+		Permissions: "0640",
+	}
+
+	// TODO: Implement adding additional files through API
+
+	cpinput := &cloudinit.ControlPlaneInput{
+		BaseUserData: cloudinit.BaseUserData{
+			AirGapped:        scope.Config.Spec.AgentConfig.AirGapped,
+			PreRKE2Commands:  scope.Config.Spec.PreRKE2Commands,
+			PostRKE2Commands: scope.Config.Spec.PostRKE2Commands,
+			ConfigFile:       initConfigFile,
+			RKE2Version:      scope.Config.Spec.AgentConfig.Version,
+			WriteFiles:       files,
+		},
+		Certificates: certificates,
+	}
+
+	cloudInitData, err := cloudinit.NewInitControlPlane(cpinput)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.storeBootstrapData(ctx, scope, cloudInitData); err != nil {
+		return ctrl.Result{}, err
+	}
 
-func Unlock(ctx context.Context, cluster *clusterv1.Cluster) bool {
-	return true
+	return ctrl.Result{}, nil
 }
 
 type RKE2InitLock interface {
-	Unlock(context.Context, *clusterv1.Cluster) bool
-	Lock(context.Context, *clusterv1.Cluster) bool
+	Unlock(ctx context.Context, cluster *clusterv1.Cluster) bool
+	Lock(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) bool
 }
 
 // TODO: Implement these functions
-func (r *Rke2ConfigReconciler) joinControlplane(ctx context.Context, scope *Scope) (res ctrl.Result, rerr error) {
+func (r *RKE2ConfigReconciler) joinControlplane(ctx context.Context, scope *Scope) (res ctrl.Result, rerr error) {
+
+	tokenSecret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: scope.Cluster.Namespace, Name: scope.Cluster.Name + "-token"}, tokenSecret); err != nil {
+		scope.Logger.Info("Token for already initialized RKE2 Cluster not found", "token-namespace", scope.Cluster.Namespace, "token-name", scope.Cluster.Name+"-token")
```

Comment:
Whats the process to fix this scenario if it happens?

Reply:
At the moment, there is no reconciliation for the token: the token is supposed to be generated and stored during cluster initialization (first control plane node). At this stage, the Reconciler supposes it exists.

We need to account for the case where it does not.

Reply:
I changed the log level to Error for now, but it definitely needs to be Reconciled.

Reply:
The Token in the case of RKE2 is very similar to the certificates in the case of Kubeadm:
- You generate it during the first init reconciliation and store it in a secret, which is your only reference.
- If a token gets deleted afterwards, let's say by an admin user, it is lost and would need a new rollout of all machines.

How that might work is another topic but:
- We could imagine that a hash of the token could be stored in an annotation to the machine while generating it, then a reconciliation section would specifically check for that annotation against the secret that would be re-generated.
- We could also imagine that we could drop a manifest or run a kubectl command as part of bootstraping the control plane nodes to store the token in a secret, which would be available in the workload cluster. Naturally, this would need the workload cluster to be available.

However, I think this should be considered in a future PR. What do you think @richardcase ?

Reply:
Create an issue to continue discussing this problem.